File: 프로그래머스/3/49191. 순위/순위.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        
        List<List<Integer>> winGraph = new ArrayList<>();
        List<List<Integer>> loseGraph = new ArrayList<>();
        
        for(int i=0; i<=n; ++i){
            winGraph.add(new ArrayList<>());
            loseGraph.add(new ArrayList<>());
        }
        
        for(int[] res: results){
            winGraph.get(res[0]).add(res[1]);
            loseGraph.get(res[1]).add(res[0]);
        }
        
        for(int i=1; i<=n; ++i){
            boolean[] visited = new boolean[n+1];
            visited[0] = true;
            visited[i] = true;
            
            Deque<Integer> queue = new ArrayDeque<>();
            queue.offerLast(i);
            
            // System.out.println("INDEX: " + i);
            
            while(!queue.isEmpty()){
                int curr = queue.pollFirst();
                visited[curr] = true;
                
                // System.out.println(curr);
                
                for(int next: winGraph.get(curr)){
                    if(visited[next] == false){
                        queue.add(next);
                    }
                }
            }
            
            queue.offerLast(i);
            
            while(!queue.isEmpty()){
                int curr = queue.pollFirst();
                visited[curr] = true;
                
                // System.out.println(curr);
                
                for(int next: loseGraph.get(curr)){
                    if(visited[next] == false){
                        queue.add(next);
                    }
                }
            }
            
            boolean isCertain = true;
            for(boolean b: visited){
                if(!b) {
                    isCertain = false;
                    break;
                }
            }
            
            if(isCertain){
                answer++;
            }
            
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49191. 순위/순위.py
from collections import defaultdict, deque

def solution(n, results):
    answer = 0
    
    win_graph = defaultdict(list)
    lose_graph = defaultdict(list)
    
    for a, b in results:
        win_graph[a].append(b)
        lose_graph[b].append(a)
    
    for i in range(1, n+1):
        visited = [False for _ in range(n+1)]
        visited[0] = True
        visited[i] = True
        
        queue = deque()
        queue.append(i)
        
        while queue:
            curr = queue.popleft()
            visited[curr] = True
            
            for next in win_graph[curr]:
                if visited[next] == False:
                    queue.append(next)
        
        queue.append(i)
        
        while queue:
            curr = queue.popleft()
            visited[curr] = True
            
            for next in lose_graph[curr]:
                if visited[next] == False:
                    queue.append(next)
        
        is_certain = True
        for v in visited:
            if v == False:
                is_certain = False
                break
                
        if is_certain: answer += 1
        
        
    return answer